test/object-store/sync/session/connection_change_notifications.cpp

```diff
@@ -64,20 +72,13 @@ TEST_CASE("sync: Connection state changes", "[sync][session][connection change]"
             return sessions_are_connected(*session);
         });
 
-        std::atomic<bool> listener1_called(false);
-        std::atomic<bool> listener2_called(false);
-        auto token1 = session->register_connection_change_callback(
-            [&](SyncSession::ConnectionState, SyncSession::ConnectionState) {
-                listener1_called = true;
-            });
-        session->unregister_connection_change_callback(token1);
```

Comment:
I'd guess that the race is because there is spurious network connection changes between the calls to register and unregister. If that is the case, what you could do is get the count of times called after unregistering the callback, and check that it is the same after calling log_out() below.